### ISSUE TITLE
Voter causes an infinite redirect for non-UsernamePasswordToken tokens

### ISSUE THREAD

**Issue by zerkms** [scheb/two-factor-bundle#62]
If you configure a project with `RememberMeToken` then when user is re-authenticated using the `RememberMe` cookie - voter's `return VoterInterface::ACCESS_DENIED;` turn the authentication process into an infinite loop.

It happens because:
1. `RememberMe` listener authenticates a user
2. On voter returns `ACCESS_DENIED`
3. Symfony's `Symfony\Component\Security\Http\Firewall\ExceptionListener` exception handler checks the user is not fully authenticated and performs a redirect (through `Symfony\Component\Security\Http\EntryPoint\FormAuthenticationEntryPoint::start()`
4. Go to step 2

Temporary solution: disable Voter

Permanent solution: the Voter implementation (and presence at all) should be reconsidered.

**Comment by umpirsky** [comment]
@zerkms I cannot reproduce it, my configuration:

```
scheb_two_factor:
    google:
        enabled: true
    trusted_computer:
        enabled: true
        cookie_secure: true
    security_tokens:
        - Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken
        - Symfony\Component\Security\Core\Authentication\Token\RememberMeToken

```

Do you have same config?

**Comment by zerkms** [comment]
What exactly did you do to reproduce it?

**Comment by umpirsky** [comment]
@zerkms Configured as above, logged in with remember me option, deleted session cookie, refreshed the page.

**Comment by zerkms** [comment]
https://github.com/zerkms/two-factor-bundle-infinite-bug-repro

Here is a repository with entirely configured application to reproduce the bug.

1. Restore composer dependencies
2. Request the `/` url
3. Put the login `foo` and password `bar` into the authentication form
4. Here you must see the two-factor-bundle form that requests a code
5. Now remove a session cookie (but leave the remember me cookie)
6. Now it fails with `localhost redirected you too many times.`

**Comment by umpirsky** [comment]
cc @scheb ?

**Comment by scheb** [comment]
Thanks for setting that up. I'll have a look as soon as I find some time. @umpirsky, did you try it yet?

**Comment by umpirsky** [comment]
@scheb Thanks for quick response. Nope, but it does not happen in my project!

**Comment by zerkms** [comment]
Here is a configured live demo http://two-factor-bundle-infinite-bug-zerkms.c9users.io/web

@umpirsky my bet is that you have not added 

    access_decision_manager:
        strategy: unanimous

https://github.com/zerkms/two-factor-bundle-infinite-bug-repro/blob/master/app/config/security.yml#L13

**Comment by umpirsky** [comment]
@zerkms Invalid credentials (foo:bar).

**Comment by zerkms** [comment]
> Invalid credentials (foo:bar).

WAAAAAAAAAAAT o_O

https://two-factor-bundle-infinite-bug-zerkms.c9users.io/web/ - `login: foo, password: bar` - those are correct credentials.

**Comment by umpirsky** [comment]
@zerkms OK, works now. But it prompts for 2FA code, and I guess you setup the account, so without your phone, we cannot test. :)

**Comment by zerkms** [comment]
@umpirsky you don't need to type 2FA code. When you see a request for code - remove the session cookie and then you get an infinite redirect loop.

**Comment by umpirsky** [comment]
@zerkms Reproduced. Does same happens after successful 2FA auth?

**Comment by zerkms** [comment]
Nope. When it is successful - the voter returns access granted.

**Comment by umpirsky** [comment]
OK, that's why I didn't reproduce it probably, will check again in my project. Thanks!

**Comment by zerkms** [comment]
So from the real world use cases when it can be harmful:

you as a user have a remember me token and an active session, but the app does not have the trusted computers feature implemented or you have not clicked a checkbox that it's a trusted computer.

Then, when your session naturally expires and you revisit the site again - the remember-me auth flow is used and you become locked with an infinite redirect. For 99% of non-developer users it is fatal - since they don't know how to clear cookies (or use anonymous mode)

**Comment by umpirsky** [comment]
I reproduced same issue in my project, this is a critical issue and I would like to get it fixed ASAP.

@scheb How can I help?

**Comment by zerkms** [comment]
@umpirsky well, you will not like my answer :-)

I think that adding a voter was a **terrible** decision. Here is why:

1. The whole idea of voters is about *who* not *how*. It's about the identity of the authenticated user, not about how they authenticated
2. The default Symfony.Security authentication implementation only provides the 3 aspects that answer on *how* question: `IS_AUTHENTICATED_FULLY`, `IS_AUTHENTICATED_REMEMBERED`, `IS_AUTHENTICATED_ANONYMOUSLY`. Whereas what the voter implementation intended to deliver is something like `IS_AUTHENTICATED_WITH_2FA`. And a proper solution would be to add it there, without voters at all. (the problem is though - the task would not be that easy, but still possible).

So my final conclusion: it must be removed, because it is simply the wrong tool chosen to solve the problem it is supposed to solve. And I personally think that trying to fix it will make it even worse.

**Comment by umpirsky** [comment]
@zerkms I see your point, but, if you disabled it entirely, how do you manage 2FA? I mean, without this voter, user never reach 2FA auth page.

**Comment by zerkms** [comment]
@umpirsky it is managed without voter entirely: by a `KernelEvents::REQUEST` event listener

https://github.com/scheb/two-factor-bundle/blob/master/Security/TwoFactor/EventListener/RequestListener.php

**Comment by umpirsky** [comment]
@zerkms Huh, so it can be removed, right? Are there any disadvantages of removing it? /cc @scheb

**Comment by zerkms** [comment]
@umpirsky well, if you check the PR it was suggested https://github.com/scheb/two-factor-bundle/pull/32

So the aim was to implement it for "userland" use.

Technically it's a breaking change, practically I don't think it was a good idea to implement it/rely on it in the very first place.

**Comment by umpirsky** [comment]
@zerkms So, it makes sense only when you have areas of the site that don't require 2FA. If all secured urls also require 2FA, you don't need a voter. Right?

**Comment by zerkms** [comment]
Well, technically if you use `is_granted` or anything similar - it may make some difference.

But in general - I agree and I cannot think of any real world use case for it.

**Comment by umpirsky** [comment]
@zerkms As far as I can tell, the difference can be, as an example:

You have:

```
exclude_pattern: /non2faarea
```

And on `/non2faarea/somepage` you use `is_granted('IS_AUTHENTICATED_FULLY')` to check if you want to render url to some 2FA protected page, it will return `true` without the voter.

**Comment by scheb** [comment]
The whole purpose of the voter is to prevent the user from getting access to "secured" content, without having finished the two-factor authentication.

Since the bundle not part of Symfony's security layer, but built on top of it - it interfers any request when the user has not completed 2FA yet - users are already fully authenticated from the perspective of Symfony's security layer. The voter is there so that the 2FA status is taken into account.

The `exclude_pattern` option was actually only meant to pass assets, which are not delivered statically, but through an controller (e.g. Assetic), and would therefore be interferred by the bundle. It was NOT meant to make any real content available, when 2FA is not completed yet. The whole concept of the bundle is to force the user to either complete 2FA or cancel (logout) it. The user should not be able to browse the website in a "halfway authenticated state". A "non2faarea" like seen above is outside the scope of the bundle.

In general, the problems arise from the fact, that the bundle it not part of the security layer - I lend the concept from how Sonata bundles did it. So far I didn't find a better way, also because I'm not that deep into Symfony's security layer. So if anyone would know how to do it properly, that would be highly welcome. But so far I haven't seen any information on how to have a "multiple step authentication process" in Symfony.

**Comment by umpirsky** [comment]
@scheb Got it, you cleared it up very well, thanks.
But we have a burning issue here, do you have any idea how to fix it?

**Comment by zerkms** [comment]
> The whole purpose of the voter is to prevent the user from getting access to "secured" content, without having finished the two-factor authentication.

That's not true. You don't need a voter to do that. The bundle worked without any problems before the PR #32.

> The voter is there so that the 2FA status is taken into account.

That's not correct either.

> But so far I haven't seen any information on how to have a "multiple step authentication process" in Symfony.

Just dig into symfony.security and see how it implements what I referred to at https://github.com/scheb/two-factor-bundle/issues/62#issuecomment-239046272

**Comment by scheb** [comment]
>> The whole purpose of the voter is to prevent the user from getting access to "secured" content, without having finished the two-factor authentication.
> That's not true. You don't need a voter to do that. The bundle worked without any problems before the PR #32.

Well, it is not that black and white as you say. There are use-cases where the voter does the right thing, e.g. when you check permissions to determine if some logic should be executed (doesn't need to be something, that becomes visible to the user on the page). This fragment of code could already been triggered, when the user has not yet completed 2FA. So there are use-cases where the voter does the right thing, but there are also ones, where it doesn't.

We can agree that the current solution has it's design flaws, but I cannot agree on simply removing the voter for good, because that would create a potential security risk for everyone using the bundle at the moment. Since you seem to know what you're doing in your application, feel free to disable the voter and take care of checking the 2FA state somehow when code relies on authentication state.

Until someone has a working solution - I don't have one currently - the voter will have to be kept in the bundle.

>> But so far I haven't seen any information on how to have a "multiple step authentication process" in Symfony.
> Just dig into symfony.security and see how it implements what I referred to at #62 (comment)

Let's say we're getting that special "IS_AUTHENTICATED_WITH_2FA" role somehow into the symfony security layer, that wouldn't change the security token already having a an authenticated user.

Here is an idea how it could work. It makes sure that there is only an authenticated token with a user object, when 2FA has been completed:

1) Login with authentication form (as an example, or any other mechanism to identify the user). Authentication listener creates unauthenticated token first, passes it to authenticator class, which gets the user from the user provider, returns the authenticated token with the user object contained.
2) Normally at this point the authenticated token would be written to token storage and an authentication success event would be fired of. At this point the bundle would need to interfer. It would need to keep the token from being written to token storage and keep the event from being fired. Instead, the authenticated token would be wrapped into a unauthenticated TwoFactorToken and that one is written to token storage. (This whole section is the tricky part where I'm not sure how this would work*)
3) A two-factor authentication listener sees the unauthenticated TwoFactorToken in the token storage and triggers based on that. It redirects user to the two-factor form (would need to be some route that has to be configured in the application, like `/login` for the login form).
4) User enters two-factor code. Two-factor authentication listener kicks in again, because the unauthenticated TwoFactorToken is still in the token storage, but this time a code has been passed and it calls the two-factor authenticator to verify the code. Code is either invalid (redirect back to two-factor authentication form with error message) or valid (the unauthenticated TwoFactorToken is unwrapped and the original authenticated user token is returned).
5) Two-factor step was successful, the authenticated user token is returned by the authenticator and is written to token storage.

* I guess we would need to interfer here https://github.com/symfony/symfony/blob/f29d46f29b91ea5c30699cf6bdb8e65545d1dd26/src/Symfony/Component/Security/Core/Authentication/AuthenticationProviderManager.php
after the line $result = $provider->authenticate($token). Decorating all objects in $this->providers could be a solution, but then the event would still be fired.

**Comment by zerkms** [comment]
> Until someone has a working solution - I don't have one currently - the voter will have to be kept in the bundle.

And as we can see - everyone simply disables it, since it creates more problems than solves :-) Not to say that the voter breaks the `exclude_pattern` option. At the moment unless you disable the voter - it simply is not taken into account.

> Let's say we're getting that special "IS_AUTHENTICATED_WITH_2FA" role somehow into the symfony security layer, that wouldn't change the security token already having a an authenticated user.

If you have that - you would have a different token or a token that implements an interface that let's you know whether user has completed 2FA or not.

The 1-5 steps look good to me, but I'm not a symfony.securitybundle expert, so I cannot tell whether it can be implemented like that.

**Comment by darookee** [comment]
As I proposed the voter I'd thought I'll chime in here. I'm not a symfony.security expert either but implementing that feature using a voter seemed a good idea at that time.

I did take a look into the symfony security-core again to see how symfony handles the authorization/authentication and it seems it uses a [voter](https://github.com/symfony/security-core/blob/master/Authorization/Voter/AuthenticatedVoter.php). That voter relies on the [AuthenticationTrustResolver](https://github.com/symfony/security-core/blob/master/Authentication/AuthenticationTrustResolver.php) and determines if the user is IS_AUTHENTICATED_FULLY when it is neither [IS_AUTHENTICATED_REMEMBERED nor IS_AUTHENTICATED_ANONYMOUSLY](https://github.com/symfony/security-core/blob/master/Authentication/AuthenticationTrustResolver.php#L71). I think the problem could be solved with extending the AuthenticationTrustResolver to check for a third condition (2FA done) and let the voter check for the *_2FA state.

**Comment by scheb** [comment]
This issue is solved in the upcoming 3.0 version. Check out the beta.
